zephyr/lib/regions_mm.c

```diff
@@ -1,54 +1,574 @@
+
+fail:
+	if (heap->span)
+		sys_mem_blocks_free_contiguous(
+		    heap->physical_blocks_allocators[mem_block_iterator], ptr,
+		    block_count);
+	else
+		sys_mem_blocks_free(
+		    heap->physical_blocks_allocators[mem_block_iterator],
+		    block_count, &ptr);
+	rfree(new_allocation);
+	return NULL;
+}
+
+/**
+ * @brief Free virtual memory heap
+ *
+ * Free the virtual memory heap object and its child allocations
+ * @param heap Ptr to the heap that is supposed to be freed
+ *
+ * @retval 0 on success;
+ * @retval -ENOTEMPTY on heap having active allocations.
+ */
+int vmh_free_heap(struct virtual_memory_heap *heap)
+{
+	if (!list_is_empty(&heap->active_allocations))
+		return -ENOTEMPTY;
+
+	for (int i = 0; i < MAX_MEMORY_ALLOCATORS_COUNT; i++) {
+		if (heap->physical_blocks_allocators[i] != 0) {
+			rfree(heap->physical_blocks_allocators[i]
+				  ->bitmap->bundles);
+			rfree(heap->physical_blocks_allocators[i]->bitmap);
+			rfree(heap->physical_blocks_allocators[i]);
+		}
+	}
+	rfree(heap);
+	return 0;
+}
+
+/**
+ * @brief Free ptr allocated on given heap
+ *
+ * Free the ptr allocation. After free action is complete
+ * check if any physical memory block can be freed up as
+ * result.
+ * @param heap Ptr to the heap that is supposed to be freed
+ * @param ptr pointer that should be freed
+ *
+ * @retval 0 on success;
+ * @retval -ENOTEMPTY on heap having active allocations.
+ */
+int vmh_free(struct virtual_memory_heap *heap, void *ptr)
+{
+	struct list_item *iterator;
+	struct vmh_active_allocation *allocation = NULL;
+	int retval;
+
+	list_for_item(iterator, &heap->active_allocations) {
+		if (ptr ==
+		    CONTAINER_OF(iterator, struct vmh_active_allocation, node)
+			->allocation_ptr) {
+
+			allocation = CONTAINER_OF(
+			    iterator, struct vmh_active_allocation, node);
 			break;
-		case MEM_REG_ATTR_OPPORTUNISTIC_MEMORY:
-			vm_heaps[i].memory_caps = SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_HP;
+		}
+	}
+
+	if (allocation == NULL)
+		return -ENOENT;
+	/* Find the allocator that hosts the allocation
+	 * and free when it is found
+	 */
+	for (int i = 0; i < MAX_MEMORY_ALLOCATORS_COUNT; i++) {
+		if (vmh_is_ptr_in_memory_range(
+			ptr, heap->physical_blocks_allocators[i]->buffer,
+			heap->physical_blocks_allocators[i]->num_blocks *
+			    (heap->physical_blocks_allocators[i]->blk_sz_shift
+			     << 1))) {
+			if (heap->span) {
+				retval = sys_mem_blocks_free_contiguous(
+				    heap->physical_blocks_allocators[i], ptr,
+				    allocation->allocation_size);
+			} else {
+				retval = sys_mem_blocks_free(
+				    heap->physical_blocks_allocators[i],
+				    allocation->allocation_size, &ptr);
+			}
 			break;
-		default:
-			return -EINVAL;
 		}
 	}
 
+	list_item_del(&allocation->node);
+
+	/* Scan for changes in physical pages mapping and
+	 * apply unmap if possible
+	 */
+	struct list_item *iterator_phys;
+	struct list_item *tmp;
+	struct list_item *iterator_activ;
+	struct vmh_active_physical_mapping *physical_mAPIng;
```

Comment:
?

Reply:
Yea , offshoot when changed api in the file to upper case ...